src/psyclone/gocean1p0.py

```diff
@@ -1879,113 +1879,177 @@ def data_on_device(self, parent):
+        :return: OpenCL iteration indices definition and initialisation.
+        :rtype: str
+        '''
+        self._check_gocean_conformity()
+
+        code = ""
+        for index, symbol in enumerate(self.argument_list[:2]):
+            code = code + self.indent(indent) + "int " + symbol.name
+            code = code + " = get_global_id(" + str(index) + ");\n"
+        return code
+
+    def gen_ocl_array_length(self, indent=0):
+        '''
+        Generate a <name>LEN<DIM> variable for each array dimension.
+        In OpenCL the sizes are retrived from the kernel global_work_size
+        (e.g. "int arrayLEN1 = get_global_size(1);")
+
+        :return: OpenCL array length variables definition and initialisation.
+        :rtype: str
+        :raises GenerationError: if the array length variable name clashes \
+                with another symbol name.
+        '''
+        self._check_gocean_conformity()
+
+        code = ""
+        for symbol in self.argument_list[2:]:
             dimensions = len(symbol.shape)
             for dim in range(1, dimensions + 1):
-                code = code + self.indent(indent + 1) + "int "
+                code = code + self.indent(indent) + "int "
                 varname = symbol.name + "LEN" + str(dim)
 
                 # Check there is no clash with other variables
-                if varname in self.symbol_table:
+                if varname in self:
+                    kname = ""
+                    if self._kernel:
+                        kname = "'{0}'".format(self._kernel.name)
                     raise GenerationError(
                         "Unable to declare the variable '{0}' to store the "
-                        "length of '{1}' because the kernel '{2}' already "
+                        "length of '{1}' because the kernel {2} already "
                         "contains a symbol with the same name."
-                        "".format(varname, symbol.name, self._name))
+                        "".format(varname, symbol.name, kname))
 
                 code = code + varname + " = get_global_size("
                 code = code + str(dim - 1) + ");\n"
+        return code
 
-        # Declare OpenCL iteration indices using the names of the first 2
-        # arguments (e.g. "int i = get_global_id(0);")
-        for index, symbol in enumerate(self.symbol_table.argument_list[:2]):
-            code = code + self.indent(indent + 1) + "int " + symbol.name
-            code = code + " = get_global_id(" + str(index) + ");\n"
+
+class GOKernelSchedule(KernelSchedule):
+    '''
+    Sub-classes KernelSchedyle to provide an API-specific implementation of the
+    OpenCL generation method and the symbol_table property.
+    '''
+
+    @property
+    def symbol_table(self):
+        '''
+        :return: Table containing symbol information for the kernel.
+        :rtype: :py:class:`psyclone.gocean1p0.GOSymbolTable`
+        '''
+        self._symbol_table.__class__ = GOSymbolTable
```

Comment:
You are right, the constructor way is nicer, I had to subclass a  GOFparser2ASTProcessor to generate a GOKernelSchedule that instantiates the GOSymbolTable.
Originally I was trying to avoid all this sub-classes, but I think it is the right way to do it.

Reply:
Yes, it does feel a bit heavyweight but it is at least clear what's going on. Thanks for improving.